v2/internal/controllers/crd_machinelearningservices_test.go

```diff
@@ -0,0 +1,264 @@
+
+				VirtualMachine: &machinelearningservices.Compute_VirtualMachine{
+					ComputeLocation:   tc.AzureRegion,
+					ComputeType:       &computeType,
+					DisableLocalAuth:  to.BoolPtr(true),
+					ResourceReference: tc.MakeReferenceFromResource(vm),
+					Properties: &machinelearningservices.VirtualMachineProperties{
+						AdministratorAccount: &machinelearningservices.VirtualMachineSshCredentials{
+							Password: to.StringPtr(password),
+							Username: to.StringPtr("bloom"),
+						},
+						SshPort: to.IntPtr(22),
+					},
+				},
+			},
+		},
+	}
+	return wsCompute
+}
+
+func newVMNetworkInterfaceWithPublicIP(tc *testcommon.KubePerTestContext, owner *genruntime.KnownResourceReference, subnet *network.VirtualNetworksSubnet, publicIP *network.PublicIPAddress, nsg *network.NetworkSecurityGroup) *network.NetworkInterface {
+
+	dynamic := network.NetworkInterfaceIPConfigurationPropertiesFormatPrivateIPAllocationMethodDynamic
+	return &network.NetworkInterface{
+		ObjectMeta: tc.MakeObjectMeta("nic"),
+		Spec: network.NetworkInterfaces_Spec{
+			Owner:    owner,
+			Location: tc.AzureRegion,
+			IpConfigurations: []network.NetworkInterfaces_Spec_Properties_IpConfigurations{{
+				Name:                      to.StringPtr("ipconfig1"),
+				PrivateIPAllocationMethod: &dynamic,
+				Subnet: &network.SubResource{
+					Reference: tc.MakeReferenceFromResource(subnet),
+				},
+				PublicIPAddress: &network.SubResource{
+					Reference: tc.MakeReferenceFromResource(publicIP),
+				},
+			}},
+			NetworkSecurityGroup: &network.SubResource{
+				Reference: tc.MakeReferenceFromResource(nsg),
+			},
+		},
+	}
+}
+
+func newNSG(tc *testcommon.KubePerTestContext, owner *genruntime.KnownResourceReference) (*network.NetworkSecurityGroup, *network.NetworkSecurityGroupsSecurityRule) {
+	protocol := network.SecurityRulePropertiesFormatProtocolTcp
+	allow := network.SecurityRulePropertiesFormatAccessAllow
+	direction := network.SecurityRulePropertiesFormatDirectionInbound
+
+	// Network Security Group
+	nsg := &network.NetworkSecurityGroup{
+		ObjectMeta: tc.MakeObjectMetaWithName(tc.Namer.GenerateName("nsg")),
+		Spec: network.NetworkSecurityGroups_Spec{
+			Location: tc.AzureRegion,
+			Owner:    owner,
+		},
+	}
+
+	// Network Security Group rule
+	rule := &network.NetworkSecurityGroupsSecurityRule{
+		ObjectMeta: tc.MakeObjectMeta("rule1"),
+		Spec: network.NetworkSecurityGroupsSecurityRules_Spec{
+			Owner:                    testcommon.AsOwner(nsg),
+			Protocol:                 &protocol,
+			SourcePortRange:          to.StringPtr("*"),
+			DestinationPortRange:     to.StringPtr("22"),
+			SourceAddressPrefix:      to.StringPtr("*"),
+			DestinationAddressPrefix: to.StringPtr("*"),
+			Access:                   &allow,
+			Priority:                 to.IntPtr(101),
+			Direction:                &direction,
+			Description:              to.StringPtr("The first rule of networking is don't talk about networking"),
+		},
+	}
```

Comment:
Suggert putting this in a separate factory method, the two are only peripherally coupled.